torch/_inductor/virtualized.py

```diff
@@ -184,6 +191,42 @@ def inner(*args):
             setattr(cls, name, make_handler(format_string))
 
 
+class TrivialOpsHandler:
+    """
+    This is intended to be used as a superclass for ops handlers which do not
+    need to do forward dataflow analysis.  This always returns None for
+    any operation; the intention is you override it and have various operations
+    perform side effects.
+    """
+    def __getattr__(self, name):
+        if name == "name":
+            return "TrivialOpsHandler"
+
+        def inner(*args, **kwargs):
+            return None
+
+        return inner
+
+    def reduction(
+        self, dtype, src_dtype, reduction_type, value
+    ) -> Union[tuple[str, ...], str]:
+        num_values = reduction_num_outputs(reduction_type)
+        return [None] * num_values if num_values > 1 else r
+
+    @staticmethod
+    def masked(mask, body, other) -> str:
+        # NB: body is a nested function, so call it so we run analysis on it
+        body()
+        return None
+
+    @staticmethod
+    def indirect_indexing(index_var, size, check=True) -> sympy.Symbol:
+        # The sympy.Symbol here is meaningless, we just return something to
+        # ensure subsequent computation with sympy expressions doesn't choke
+        # (as we don't control sympy IR functions)
+        return sympy_symbol(f"({str(index_var)})")
```

Comment:
This may be problematic since IIUC our indexing expressions will now use the sympy variable `(None)` to represent all indirect indexing symbols, where before it would be `tmp{n}`. This could mean `MemoryDep`s which are incompatible due to using different indirect indexing variables would end up appearing to match.

Reply:
Hmmm. If what you say is true, then probably we should not do this refactor at all, because the string generation is actually load bearing, for indirect indexing purposes. Thinking about it some more...

Reply:
I think the unique variable names are load bearing, but we don't actually need to accumulate an output string or format the assignment line. So the default implementation could just be
```python 
def inner(*args, **kwargs):
     return f"tmp{next(self.var_counter)}"
```

Reply:
Oh my god this is so cursed

Reply:
OK, so I played around with this code experimentally, and also did some more code reading. What you have described is indeed true.  However, I find I am still confused: the code here is ostensibly trying to identify read/write dependencies for external buffers; i.e., we are going to be comparing deps across multiple DBR fragments. However, whenever we have an indirect indexing, we are only ever going to get a `tmp{n}`, where this is an allocation that is only unique to a particular `inner_fn`. So let's suppose that I have two `MemoryDep(buf0, (tmp4))` from two separate `inner_fn`s. Do these refer to the same indexing pattern? Clearly not! And yet I cannot see how we could possibly disambiguate in this case. So isn't this likely to just be wrong anyway?

Reply:
The logic here doesn't seem to test if it's indirect though:

```
    def can_inplace(self, read_dep: dependencies.MemoryDep):
        if self.get_aliases() or self.is_template():
            return False
        if len(self.read_writes.writes) == 1 and isinstance(
            read_dep, dependencies.MemoryDep
        ):  
            write_dep = next(iter(self.read_writes.writes))
            assert isinstance(write_dep, dependencies.MemoryDep), f"{type(write_dep)=}"
            return read_dep.index == write_dep.index and read_dep.size == write_dep.size
        return False
```

Reply:
IIUC the issue you describe comes up when comparing `MemoryDep`s from different `ir.Loops` instances, whereas `can_inplace` here is only called for with `MemoryDep` insteaces from the same unfused `SchedulerNode` instance. So I believe it will never run into this issue.

Reply:
Indirect indexing should never match across kernel calls, though within a single kernel it can.